path_login.go

```diff
@@ -0,0 +1,239 @@
+	}
+
+	if req.Connection != nil && !cidrutil.RemoteAddrIsOk(req.Connection.RemoteAddr, role.BoundCIDRs) {
+		return logical.ErrorResponse("request originated from invalid CIDR"), nil
+	}
+
+	config, err := b.config(ctx, req.Storage)
+	if err != nil {
+		return nil, err
+	}
+	if config == nil {
+		return logical.ErrorResponse("could not load configuration"), nil
+	}
+
+	// Here is where things diverge. If it is using OIDC discovery, validate
+	// that way; otherwise validate against the locally configured keys. Once
+	// things are validated, we re-unify the request path when evaluating the
+	// claims.
+	allClaims := map[string]interface{}{}
+	switch {
+	case len(config.ParsedJWTPubKeys) != 0:
+		parsedJWT, err := jwt.ParseSigned(token)
+		if err != nil {
+			return logical.ErrorResponse(errwrap.Wrapf("error parsing token: {{err}}", err).Error()), nil
+		}
+
+		claims := jwt.Claims{}
+
+		var valid bool
+		for _, key := range config.ParsedJWTPubKeys {
+			if err := parsedJWT.Claims(key, &claims, allClaims); err == nil {
+				valid = true
+				break
+			}
+		}
+		if !valid {
+			return logical.ErrorResponse("no known key successfully validated the token signature"), nil
+		}
+
+		expected := jwt.Expected{
+			Issuer:   config.BoundIssuer,
+			Subject:  role.BoundSubject,
+			ID:       role.BoundID,
+			Audience: jwt.Audience(role.BoundAudiences),
+			Time:     time.Now(),
+		}
+
+		if err := claims.Validate(expected); err != nil {
+			return logical.ErrorResponse(errwrap.Wrapf("error validating claims: {{err}}", err).Error()), nil
+		}
+
+	case config.OIDCIssuerURL != "":
+		provider, err := b.getProvider(ctx, config)
+		if err != nil {
+			return nil, errwrap.Wrapf("error getting provider for login operation: {{err}}", err)
+		}
+
+		verifier := provider.Verifier(&oidc.Config{
+			SkipClientIDCheck: true,
+		})
+
+		idToken, err := verifier.Verify(ctx, token)
+		if err != nil {
+			return logical.ErrorResponse(errwrap.Wrapf("error validating signature: {{err}}", err).Error()), nil
+		}
+
+		if err := idToken.Claims(allClaims); err != nil {
+			return logical.ErrorResponse(errwrap.Wrapf("unable to successfully parse all claims from token: {{err}}", err).Error()), nil
+		}
+
+		var jtiVal string
+		if jti, ok := allClaims["jti"]; ok {
+			if jtiStr, ok := jti.(string); ok {
+				jtiVal = jtiStr
+			}
+		}
+
+		switch {
+		case config.BoundIssuer != "" && config.BoundIssuer != idToken.Issuer:
+			return logical.ErrorResponse("iss claim does not match bound issuer"), nil
+		case role.BoundSubject != "" && role.BoundSubject != idToken.Subject:
+			return logical.ErrorResponse("sub claim does not match bound subject"), nil
+		case len(role.BoundAudiences) != 0 && !strutil.StrListSubset(role.BoundAudiences, idToken.Audience):
+			return logical.ErrorResponse("aud claim does not match any bound audience"), nil
+		case role.BoundID != "" && role.BoundID != jtiVal:
+			return logical.ErrorResponse("jti claim does not match bound id"), nil
+		}
```

Comment:
I think it might be cool to have a map of custom claims you want to validate agains in a role.

Reply:
Agreed, but I'd rather save that for later than do it in the initial cut.

Reply:
👍